File: tests/workflows/tinyfaas/IoT/CS/handler.js
// IoT-CS: CheckSound - calls CSL and CSA in parallel (sync)
// Express-style handler (req, res)

import axios from "axios";

const GATEWAY_BASE = process.env.TINYFAAS_GATEWAY_URL || "http://tinyfaas.com";

function callFunction(functionName, data, sync, incomingHeaders) {
    return (async () => {
        const pathPrefix = sync ? "/fn/" : "/async-fn/";
        const url = new URL(`${GATEWAY_BASE}${pathPrefix}iot-${functionName}`);

        const headers = Object.assign({}, incomingHeaders || {});
        delete headers.host;
        delete headers["content-length"];
        headers["content-type"] = "application/json";

        try {
            const res = await axios.post(url.toString(), data, {
                headers,
                timeout: 60000,
                maxRedirects: 0,
                validateStatus: () => true,
            });

            if (!sync) {
                return {};
            }

            try {
                if (res.data === undefined || res.data === null || res.data === "") {
                    return {};
                }
                if (typeof res.data === "object") {
                    return res.data;
                }
                return JSON.parse(res.data);
            } catch {
                return {};
            }
        } catch (err) {
            console.error(`Error calling ${functionName}:`, err && err.message ? err.message : err);
            throw err;
        }
    })();
}

export default async (req, res) => {
    let event = req.body;
    console.log("CheckSound: Event:", event);

    // Augment event with location and sensorID
    event.location = event.originalEvent ? event.originalEvent.sensorID : 0;
    event.sensorID = event.originalEvent ? event.originalEvent.sensorID : 0;

    // Sync parallel calls to csl and csa
    let calls = [];
    calls.push(callFunction("csl", event, true, req.headers));
    calls.push(callFunction("csa", event, true, req.headers));

    console.log("CheckSound has called Loud and Accident");

    let results = await Promise.all(calls);

    res.json({
        from: "CheckSound",
        calls: results,
    });
};

File: tests/workflows/tinyfaas/linear3/a/handler.py
import os
import logging
import requests

GATEWAY_BASE = os.environ.get("TINYFAAS_GATEWAY_URL", "http://tinyfaas.com")
logging.basicConfig(
    level=logging.INFO,
    format="{asctime} {levelname} {message}",
    style="{",
    datefmt="%Y-%m-%d %H:%M:%S",
)
logging.getLogger().setLevel(logging.INFO)


def _build_forward_headers(request) -> dict[str, str]:
    forwarded = dict(request.headers)
    forwarded.pop("Host", None)
    forwarded.pop("Content-Length", None)
    return forwarded


def handle(request):
    logging.info("Function linear3-a is called")
    logging.info("Function linear3-a is calling linear3-b")
    res = requests.post(
        f"{GATEWAY_BASE}/fn/linear3-b",
        headers=_build_forward_headers(request),
        json={},
        timeout=30,
    )
    return {
        "body": {
            "msg": "Function linear3-a is finished",
            "data": res.json(),
            "request_headers": dict(request.headers),
            "env": dict(os.environ),
        }
    }
